src/coremods/core_whois.cpp

```diff
@@ -177,10 +177,11 @@ void CommandWhois::SendChanList(WhoisContextImpl& whois)
 	{
 		Membership* memb = *i;
 		Channel* c = memb->chan;
-		/* If the target is the sender, neither +p nor +s is set, or
-		 * the channel contains the user, it is not a spy channel
+		/* If neither +p nor +s is set, or the channel contains the user
+		 * and the user is not running whois on themselves, it is not
+		 # a spy channel
 		 */
-		if ((whois.IsSelfWhois()) || ((!c->IsModeSet(privatemode)) && (!c->IsModeSet(secretmode))) || (c->HasUser(whois.GetSource())))
+		if ((!c->IsModeSet(privatemode) && !c->IsModeSet(secretmode)) || (c->HasUser(whois.GetSource()) && !whois.IsSelfWhois()))
```

Comment:
It might also be desirable to split private/secret channels when the source is not the target user but is also on the channel. What do you think?